backend/common/providers/s3_provider.py

```diff
@@ -0,0 +1,46 @@
+import logging
+
+import boto3
+
+from backend.layers.thirdparty.s3_provider_interface import S3ProviderInterface
+
+AWS_S3_MAX_ITEMS_PER_BATCH = 1000
+
+logger = logging.getLogger(__name__)
+
+
+class S3Provider(S3ProviderInterface):
```

Comment:
Is it necessary to use the S3Provider class instead of doing what the WMG and CellGuide pipelines do to upload their data (they use `subprocess` to invoke aws CLI directly)?

https://github.com/chanzuckerberg/single-cell-data-portal/blob/3ed9e0a7da5a4de26eda52a85cb8372fcb60ee32/backend/cellguide/pipeline/__init__.py#L68

Instead of uploading one file at a time, we could use `aws s3 sync` to upload the local directory with new GPT descriptions to the remote directory.

Reply:
Granted, `S3Provider.does_object_exist` is nicer than parsing captured stdout from a `subprocess` run. Could you implement a sync function in this provider and use that instead of `upload_file`?

Reply:
i think in general it's better to use a provider for a 3rd party service so that the calling service isn't familiar with the internal workings of the 3rd party service. as far as the cellguide pipeline is aware, it only needs to "upload files to storage service" or "check if object exists at this path in the storage service. a few benefits of this compartmentalization:

1. abstracting away the internal workings of a 3rd party provider makes it a bit cleaner if you ever want to switch cloud providers, say to google cloud or azure.
2. reduces code reuse. the same reason why we might want to package `output_json_per_key` into `utils.py` instead of copying and pasting the same code everywhere applies here.
3. reusing code reduces the odds of introducing bugs. in the case where you're executing the aws cli directly, you might accidentally miss a param in there or something.
4. code reviewers can have have confidence that `S3Provider.upload_file` works as expected without needing to review or understand any underlying logic backing that

Reply:
Could you implement a sync function in the provider and use that for the GPT pipeline instead of uploading files one at a time?

Reply:
updated!